src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -669,23 +690,32 @@ - (NSArray *)_getAttributesToReportWithReportedValues:(NSArray<NSDictionary<NSSt
         // if this is an error, report and purge cache
         if (attributeError) {
             shouldReportAttribute = YES;
+            MTR_LOG_INFO("MTRDevice report %@ error %@ purge expected value %@ read cache %@", attributePath, attributeError,
+                _expectedValueCache[attributePath], _readCache[attributePath]);
             _expectedValueCache[attributePath] = nil;
             _readCache[attributePath] = nil;
         } else {
-            // check write cache and purge if needed
+            // if expected values exists, purge and update read cache
             MTRPair<NSDate *, NSDictionary *> * expectedValue = _expectedValueCache[attributePath];
             if (expectedValue) {
                 if (![self _attributeDataValue:attributeDataValue isEqualToDataValue:expectedValue.second]) {
                     shouldReportAttribute = YES;
                 }
                 _expectedValueCache[attributePath] = nil;
-            }
-
-            // then compare and update read cache
-            if (![self _attributeDataValue:attributeDataValue isEqualToDataValue:_readCache[attributePath]]) {
+                _readCache[attributePath] = attributeDataValue;
+            } else if (![self _attributeDataValue:attributeDataValue isEqualToDataValue:_readCache[attributePath]]) {
```

Comment:
I'd think we want to update `_readCache` no matter what the state of `_expectedValueCache` was here...

That is, I think the logic should be:

1. Report if we have an expected value and it does not match, or if we have no expected value and the read cache value does not match.
2. Clear out the expected value cache if there is a value there.
3. Update the read cache.

Reply:
Ah, I missed the `_readCache[attributePath] = attributeDataValue;` bit in the `expectedValue` path. So this part is all good.